### ISSUE TITLE
Feature Request: support variable substitution in manifest.yml

### ISSUE THREAD

**Issue by imod** [cloudfoundry/cli#820]
I would like to reference a buildpack hosted in a private git repository as described at https://docs.cloudfoundry.org/buildpacks/custom.html#deploying-with-custom-buildpacks so I need to add username/password to the git repo url. 
But I would like to reference the buildpack in the `manifest.yml` and not on the commandline and without the username/passwords in clear text. Therefore I would like to do something like this:

```
buildpack: https://$MY_USERANAME:$MY_PASSWORD@github.com/johndoe/my-buildpack.git
```

…and set `MY_USERANAME` and `MY_PASSWORD` as environment variable (or argument) for CLI

**Comment by dkoper** [comment]
Hi @imod 

We've been reluctant to allow environment variables in app manifests, one reason being that the app manifest content itself would not actually reflect what you have just deployed.

But your use case is a valid one that we should find a way to cater for.

In the meantime, I wonder if you would consider using app manifest inheritance to address the issue:
http://docs.cloudfoundry.org/devguide/deploy-apps/manifest.html#multi-manifests

In your manifest you add `inherit: base-manifest.yml` before `applications:`, and you'd place the `buildpacks:` line with username and password in the base-manifest.yml file.
This way you can safely store your app manifest in your source versioning system and potentially script the creation of the base manifest just before the push and remove it afterwards, or store it securely elsewhere?

Regards,
Dies Koper
CF CLI PM

**Comment by imod** [comment]
@dkoper I think using manifest inheritance and generate the `manifest.yml` on the fly is a lot more complicated then just pass the full buildpack uri to CLI (with `-b`). ...and at the end its not visible to the users/devs which buildpack should be used for the given `manifest.yml` either.

**Comment by dkoper** [comment]
@imod We've discussed it in the team and are still reluctant to introduce environment variables in app manifests. It seems inheritance solves the issue of credentials in the (main) manifest, as does setting/overwriting the buildpack url using the `-b` option. You could always add a ` # comment` in your manifest for your users/devs which buildpack should be used as a shortcut to opening up the parent manifest.

**Comment by imod** [comment]
well, that's unfortunate - adding a comment has not the same effect, as in this case a user pushing the app without defining `-b` then cli will just fallback to the default buildpack and therefore the app will not work as expected. To fully follow the 12 factors - I think its a must that we are able to define the buildpack in the manifest, because its required to make the app work properly - on the other hand the passwords are not required, because they could change over time without having an impact on the runtime of the app.

**Comment by krishicks** [comment]
What about adding `buildpack_basic_auth` or similar, which takes the user and password to be constructed as a basic auth URL later?

**Comment by manjatav** [comment]
How to supply environment variable in jenkins for cloud foundry plugin.
Like target, space and organization.

**Comment by jketo** [comment]
We like to keep manifest files on the version control system and have them to be exactly the same for all the developers and all the systems. Without env substitution sharing manifests becomes a pain and a possible place for information leakage, if kept under same versioning. This limits the usability quite much.

**Comment by dross91786** [comment]
+1 on this. Being able to dynamically update would curb the need for multiple manifests based on the target env.

**Comment by turingmachine** [comment]
+1

**Comment by tjvman** [comment]
For folks interested in this feature, we have a current track of work to allow interpolating variables with a `--vars-file` flag in `cf push`. [Tracker epic is here](https://www.pivotaltracker.com/epic/show/3904246)

**Comment by abbyachau** [comment]
Hi @imod we've published the variable substitution feature in `6.37.0`; see [release notes](https://github.com/cloudfoundry/cli/releases/tag/v6.37.0) for more information. Please do not hesitate to reach out if you have feedback by creating a new issue. Thanks.
